python/prairielearn.py

```diff
@@ -52,6 +53,45 @@ class QuestionData(TypedDict):
 class ElementTestData(QuestionData):
     test_type: Literal['correct', 'incorrect', 'invalid']
 
+
+def set_weighted_score_data(data: QuestionData, weight_default: int=1) -> None:
+    """
+    Sets overall question score to be weighted average of all partial scores. Uses
+    weight_default to fill in a default weight for a score if one is missing.
+    """
+
+    weight_total = 0
+    score_total = 0.0
+    for part in data["partial_scores"].values():
+        score = part["score"]
+        weight = part.get("weight", weight_default)
+
+        if score is None:
+            raise ValueError("Can't set weighted score data if score is None.")
+
+        score_total += score * weight
+        weight_total += weight
+
+    data['score'] = score_total / weight_total
+
+def set_all_or_nothing_score_data(data: QuestionData) -> None:
+    """Gives points to main question score if all partial scores are correct."""
+
+    data['score'] = 1.0 if all_questions_correct(data) else 0.0
+
+def all_questions_correct(data: QuestionData) -> bool:
```

Comment:
Another use of the word "question" to refer to something that's not actually the question as a whole. Maybe name this `all_partial_scores_correct`?

Reply:
Renamed! I think merging this should wait until we have the linter running over these files (which should be resolved after the black PR is done).